File: demo/KSYLiveDemo/QRViewController.m
//
//  QRViewController.m
//  KSYLiveDemo
//
//  Created by 孙健 on 16/4/13.
//  Copyright © 2016年 qyvideo. All rights reserved.
//

#import "QRViewController.h"
#import <AVFoundation/AVFoundation.h>
@interface QRViewController ()<AVCaptureMetadataOutputObjectsDelegate>{
    UIView      *_viewPreview;          //预览视图
    UILabel     *_QRLabel;              //地址栏
    UIButton    *_scanBtn;              //扫描按钮
    UIView      *_boxView;              //扫面框
    BOOL        _isReading;             //正在扫描
    CALayer     *_scanLayer;            //扫描图层
    CGFloat     _width;
    CGFloat     _height;
    CGFloat     _btnWdt;
    
    UIButton    *_cancelBtn;             // 退出扫码界面
    
}
- (BOOL)startReading;
- (void)stopReading;

@property (nonatomic, strong) AVCaptureSession *captureSeesion;
@property (nonatomic, strong) AVCaptureVideoPreviewLayer  *videoPreviewLayer;

@end

@implementation QRViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    //init
    [self initVariable];
    [self addViews];
    [self startReading];
}

- (void)initVariable{
    _captureSeesion = nil;
    _isReading      = NO;
    _width  = self.view.frame.size.width;
    _height = self.view.frame.size.height;
}

- (void)addViews{
    _viewPreview = [self addViewPreview];
    _QRLabel     = [self addLable];
    _scanBtn     = [self addButton];
    _cancelBtn   = [self addCancelButton];
}

- (UIButton *)addCancelButton {
    UIButton *button = [[UIButton alloc]initWithFrame:CGRectMake(_width / 2, _height - 30, _width / 2, 30)];
    [self.view addSubview:button];
    [button setTitle:@"取消" forState:UIControlStateNormal];
    [button addTarget:self action:@selector(quitQRScaner) forControlEvents:UIControlEventTouchUpInside];
    button.layer.masksToBounds = YES;
    [button setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
    button.layer.masksToBounds = YES;
    button.layer.borderColor   = [UIColor blackColor].CGColor;
    button.layer.borderWidth   = 1;
    button.layer.cornerRadius  = 5;
    return button;
}

- (void)quitQRScaner {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (UIView *)addViewPreview{
    UIView *view = [[UIView alloc]initWithFrame:CGRectMake(0, 64, _width, _height - 94)];
    [self.view addSubview:view];
    return view;
}
- (UILabel *)addLable{
    UILabel *label = [[UILabel alloc]initWithFrame:CGRectMake(0, 64, _width, 30)];
    label.backgroundColor = [UIColor whiteColor];
    [self.view addSubview:label];
    label.layer.masksToBounds = YES;
    label.layer.borderWidth   = 1;
    label.layer.borderColor   = [UIColor blackColor].CGColor;
    label.layer.cornerRadius  = 2;
    [self.view addSubview:label];
    return label;
}
- (UIButton *)addButton{
    UIButton *button = [[UIButton alloc]initWithFrame:CGRectMake(0, _height - 30, _width / 2, 30)];
    [self.view addSubview:button];
    [button setTitle:@"正在扫描..." forState:UIControlStateNormal];
    [button addTarget:self action:@selector(reScan) forControlEvents:UIControlEventTouchUpInside];
    button.layer.masksToBounds = YES;
    [button setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
    button.layer.masksToBounds = YES;
    button.layer.borderColor   = [UIColor blackColor].CGColor;
    button.layer.borderWidth   = 1;
    button.layer.cornerRadius  = 5;
    return button;
}
- (void)reScan{
    if (!_isReading) {
        if ([self startReading]) {
            [_scanBtn setTitle:@"正在扫描..." forState:UIControlStateNormal];
            [_QRLabel setText:@"Scanning for QR Code"];
            
        }
    }
    else{
        [self stopReading];
        [_scanBtn setTitle:@"重新扫描" forState:UIControlStateNormal];
    }
    
    _isReading = !_isReading;
}
// start reading
- (BOOL)startReading{
    NSError *error;
    
    //1.初始化捕捉设备（AVCaptureDevice），类型为AVMediaTypeVideo
    AVCaptureDevice *captureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    //2.用captureDevice创建输入流
    AVCaptureDeviceInput *input = [AVCaptureDeviceInput deviceInputWithDevice:captureDevice error:&error];
    if (!input) {
        NSLog(@"QRVC: %@", [error localizedDescription]);
        return NO;
    }
    
    //3.创建媒体数据输出流
    AVCaptureMetadataOutput *captureMetadataOutput = [[AVCaptureMetadataOutput alloc] init];
    
    //4.实例化捕捉会话
    _captureSeesion = [[AVCaptureSession alloc] init];
    
    //4.1.将输入流添加到会话
    [_captureSeesion addInput:input];
    
    //4.2.将媒体输出流添加到会话中
    [_captureSeesion addOutput:captureMetadataOutput];
    
    //5.创建串行队列，并加媒体输出流添加到队列当中
    dispatch_queue_t dispatchQueue;
    dispatchQueue = dispatch_queue_create("myQueue", NULL);
    //5.1.设置代理
    [captureMetadataOutput setMetadataObjectsDelegate:self queue:dispatchQueue];
    
    //5.2.设置输出媒体数据类型为QRCode
    [captureMetadataOutput setMetadataObjectTypes:[NSArray arrayWithObject:AVMetadataObjectTypeQRCode]];
    
    //6.实例化预览图层
    _videoPreviewLayer = [[AVCaptureVideoPreviewLayer alloc] initWithSession:_captureSeesion];
    
    //7.设置预览图层填充方式
    [_videoPreviewLayer setVideoGravity:AVLayerVideoGravityResizeAspectFill];
    
    //8.设置图层的frame
    [_videoPreviewLayer setFrame:_viewPreview.layer.bounds];
    
    //9.将图层添加到预览view的图层上
    [_viewPreview.layer addSublayer:_videoPreviewLayer];
    
    //10.设置扫描范围
    captureMetadataOutput.rectOfInterest = CGRectMake(0.2f, 0.2f, 0.8f, 0.8f);
    
    //10.1.扫描框
    _boxView = [[UIView alloc] initWithFrame:CGRectMake(_viewPreview.bounds.size.width * 0.2f, _viewPreview.bounds.size.height * 0.2f, _viewPreview.bounds.size.width - _viewPreview.bounds.size.width * 0.4f, _viewPreview.bounds.size.height - _viewPreview.bounds.size.height * 0.4f)];
    _boxView.layer.borderColor = [UIColor greenColor].CGColor;
    _boxView.layer.borderWidth = 1.0f;
    
    [_viewPreview addSubview:_boxView];
    
    //10.2.扫描线
    _scanLayer = [[CALayer alloc] init];
    _scanLayer.frame = CGRectMake(0, 0, _boxView.bounds.size.width, 1);
    _scanLayer.backgroundColor = [UIColor brownColor].CGColor;
    
    [_boxView.layer addSublayer:_scanLayer];
    
    NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:0.2f target:self selector:@selector(moveScanLayer:) userInfo:nil repeats:YES];
    
    [timer fire];
    
    //10.开始扫描
    [_captureSeesion startRunning];
    
    return YES;
}
- (void)moveScanLayer:(NSTimer *)timer
{
    CGRect frame = _scanLayer.frame;
    if (_boxView.frame.size.height < _scanLayer.frame.origin.y) {
        frame.origin.y = 0;
        _scanLayer.frame = frame;
    }else{
        
        frame.origin.y += 5;
        
        [UIView animateWithDuration:0.1 animations:^{
            _scanLayer.frame = frame;
        }];
    }
}
#pragma mark - AVCaptureMetadataOutputObjectsDelegate
- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection
{
    //判断是否有数据
    if (metadataObjects != nil && [metadataObjects count] > 0) {
        AVMetadataMachineReadableCodeObject *metadataObj = [metadataObjects objectAtIndex:0];
        //判断回传的数据类型
        if ([[metadataObj type] isEqualToString:AVMetadataObjectTypeQRCode]) {
            [_QRLabel performSelectorOnMainThread:@selector(setText:) withObject:[metadataObj stringValue] waitUntilDone:NO];
            [self performSelectorOnMainThread:@selector(stopReading) withObject:nil waitUntilDone:NO];
            _isReading = NO;
        }
    }
}
-(void)stopReading{
    [_captureSeesion stopRunning];
    _captureSeesion = nil;
    [_scanLayer removeFromSuperlayer];
    [_videoPreviewLayer removeFromSuperlayer];
    [_scanBtn setTitle:@"重新扫描" forState:UIControlStateNormal];
    if (self.getQrCode) {
        self.getQrCode(_QRLabel.text);
    }
    [self.navigationController popViewControllerAnimated:YES];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: demo/KSYLiveDemo_Swift/KSYLiveDemo_Swift/QRViewController.swift
//
//  QRViewController.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/13.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

class QRViewController: UIViewController {

    var getQrCode: ((_ stringQR: String) -> Void)?
    
    private
    var _viewPreview: UIView?
    var _QRLabel: UILabel?
    var _scanBtn: UIButton?
    var _backBtn: UIButton?
    var _boxView: UIView?
    var _isReading: Bool = false
    var _scanLayer: CALayer?
    var _width: CGFloat = 0
    var _height: CGFloat = 0
    
    var captureSession: AVCaptureSession?
    var videoPreviewLayer: AVCaptureVideoPreviewLayer?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .white
        
        initVariable()
        addViews()
        startReading()
    }

    func initVariable() {
        captureSession = nil
        _isReading = false
        _width = view.frame.width
        _height = view.frame.height
    }
    
    func addViews() {
        _viewPreview = addViewPreview()
        _QRLabel = addLable()
        _scanBtn = addButton(title: "正在扫描...")
        _backBtn = addButton(title: "返回")
        _scanBtn?.frame = CGRect.init(x: 0,
                                      y: _height - 30,
                                      width: _width,
                                      height: 30)
        _backBtn?.frame = CGRect.init(x: 20,
                                      y: 30,
                                      width: 80,
                                      height: 30)
    }
    
    func addViewPreview() -> UIView {
        let view = UIView.init(frame: CGRect.init(x: 0, y: 64, width: _width, height: _height - 94))
        self.view.addSubview(view)
        return view
    }
    
    func addLable() -> UILabel {
        let label = UILabel.init(frame: CGRect.init(x: 0, y: 64, width: _width, height: 30))
        label.backgroundColor = .white
        view.addSubview(label)
        label.layer.masksToBounds = true
        label.layer.borderWidth = 1
        label.layer.borderColor = UIColor.black.cgColor
        label.layer.cornerRadius = 2
        view.addSubview(label)
        return label
    }
    
    func addButton(title: String) -> UIButton {
        let btn = UIButton.init(type: .roundedRect)
        btn.setTitle(title, for: .normal)
        btn.backgroundColor = .lightGray
        view.addSubview(btn)
        btn.addTarget(self, action: #selector(onBtn(sender:)), for: .touchUpInside)
        return btn
    }
    
    func onBtn(sender: UIButton){
        if sender == _scanBtn {
            reScan()
        }else if sender == _backBtn {
            dismiss(animated: false, completion: nil)
        }
    }
    
    func reScan() {
        if !_isReading {
            if startReading() {
                _scanBtn?.setTitle("正在扫描...", for: .normal)
                _QRLabel?.text = "Scanning for QR Code"
            }
        }else{
            stopReading()
            _scanBtn?.setTitle("重新扫描", for: .normal)
        }
        _isReading = !_isReading
    }
    
    func startReading() -> Bool {
        // 1.初始化捕捉设备（AVCaptureDevice），类型为AVMediaTypeVideo
        let captureDevice = AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo)
        // 2.用captureDevice创建输入流
        var input: AVCaptureDeviceInput? = nil
        do{
            input = try AVCaptureDeviceInput.init(device: captureDevice)
        }catch let error as NSError{
            print("\(error.debugDescription)")
        }
        
        guard let _ = input else {
            return false
        }

        // 3.创建媒体数据输出流
        let captureMetadataOutput = AVCaptureMetadataOutput()
        // 4.实例化捕捉会话
        captureSession = AVCaptureSession()
        // 4.1 将输入流添加到会话
        captureSession?.addInput(input!)
        // 4.2 将媒体输出流添加到会话中
        captureSession?.addOutput(captureMetadataOutput)
        // 5.创建串行队列，并加媒体输出流添加到队列当中
        let outputQueue = DispatchQueue.init(label: "outputQueue")
        // 5.1.设置代理
        captureMetadataOutput.setMetadataObjectsDelegate(self, queue: outputQueue)
        // 5.2.设置输出媒体数据类型为QRCode
        captureMetadataOutput.metadataObjectTypes = [AVMetadataObjectTypeQRCode]
        
        // 6.实例化预览图层
        videoPreviewLayer = AVCaptureVideoPreviewLayer.init(session: captureSession)
        // 7.设置预览图层填充方式
        videoPreviewLayer?.videoGravity = AVLayerVideoGravityResizeAspectFill
        // 8.设置图层的frame
        videoPreviewLayer?.frame = _viewPreview!.layer.bounds
        
        // 9.将图层添加到预览view的图层上
        _viewPreview?.layer.addSublayer(videoPreviewLayer!)
        
        // 10.设置扫描范围
        captureMetadataOutput.rectOfInterest = CGRect.init(x: 0.2, y: 0.2, width: 0.8, height: 0.8)
        // 10.1.扫描框
        _boxView = UIView.init(frame: CGRect.init(x: _viewPreview!.bounds.size.width * 0.2,
                                                  y: _viewPreview!.bounds.size.height * 0.2,
                                                  width: _viewPreview!.bounds.size.width - _viewPreview!.bounds.size.width * 0.4,
                                                  height: _viewPreview!.bounds.size.height - _viewPreview!.bounds.size.height * 0.4))
        _boxView?.layer.borderColor = UIColor.green.cgColor
        _boxView?.layer.borderWidth = 1.0
        
        _viewPreview?.addSubview(_boxView!)
        
        // 10.2.扫描线
        _scanLayer = CALayer()
        _scanLayer?.frame = CGRect.init(x: 0, y: 0, width: _boxView!.frame.width, height: 1)
        _scanLayer?.backgroundColor = UIColor.brown.cgColor
        
        _boxView?.layer.addSublayer(_scanLayer!)
        
        let timer = Timer.scheduledTimer(timeInterval: 0.2, target: self, selector: #selector(moveScanLayer(timer:)), userInfo: nil, repeats: true)
        
        timer.fire()
        
        // 11.开始扫描
        captureSession?.startRunning()
        
        return true
    }
    
    func stopReading() {
        captureSession?.stopRunning()
        captureSession = nil
        _scanLayer?.removeFromSuperlayer()
        videoPreviewLayer?.removeFromSuperlayer()
        _scanBtn?.setTitle("重新扫描", for: .normal)
        if let _ = getQrCode {
            getQrCode!(_QRLabel!.text ?? "")
        }
    }

    func moveScanLayer(timer: Timer) {
        var frame = _scanLayer!.frame
        if _boxView!.frame.height < _scanLayer!.frame.origin.y {
            frame.origin.y = 0
            _scanLayer?.frame = frame
        }else {
            frame.origin.y += 5
            UIView.animate(withDuration: 0.1, animations: { [weak self] in
                self?._scanLayer?.frame = frame
            })
        }
    }
}

extension QRViewController: AVCaptureMetadataOutputObjectsDelegate {
    func captureOutput(_ captureOutput: AVCaptureOutput!, didOutputMetadataObjects metadataObjects: [Any]!, from connection: AVCaptureConnection!) {
        // 判断是否有数据
        if metadataObjects != nil && metadataObjects.count > 0 {
            let metadataObj: AVMetadataMachineReadableCodeObject = metadataObjects.first as! AVMetadataMachineReadableCodeObject
            //判断回传的数据类型
            if metadataObj.type == AVMetadataObjectTypeQRCode {
                DispatchQueue.main.async { [weak self] in
                    self?._QRLabel?.text = metadataObj.stringValue
                }
                performSelector(onMainThread: #selector(stopReading), with: nil, waitUntilDone: false)
                _isReading = false
            }
        }
    }
}
